cirq/ops/raw_types.py

```diff
@@ -445,6 +464,143 @@ def validate_args(self, qubits: Sequence['cirq.Qid']):
         _validate_qid_shape(self, qubits)
 
 
+@value.value_equality
+class TaggedOperation(Operation):
+    """A specific operation instance that has been identified with a set
+    of Tags for special processing.  This can be initialized with
+    Using Operation.with_tags(tag) or by TaggedOperation(op, tag).
+
+    See Operation.with_tags() for more information on intended usage.
+    """
+
+    def __init__(self, sub_operation: 'cirq.Operation', *tags: Any):
+        self.sub_operation = sub_operation
+        self._tags = list(tags)
+
+    @property
+    def qubits(self) -> Tuple['cirq.Qid', ...]:
+        return self.sub_operation.qubits
+
+    @property
+    def gate(self) -> Optional['cirq.Gate']:
+        return self.sub_operation.gate
+
+    def with_qubits(self, *new_qubits: 'cirq.Qid'):
+        return TaggedOperation(self.sub_operation.with_qubits(*new_qubits),
+                               *self._tags)
+
+    def controlled_by(self,
+                      *control_qubits: 'cirq.Qid',
+                      control_values: Optional[Sequence[
+                          Union[int, Collection[int]]]] = None
+                     ) -> 'cirq.Operation':
+        return self.sub_operation.controlled_by(*control_qubits,
+                                                control_values=control_values)
+
+    @property
+    def tags(self):
+        return self._tags
+
+    def with_tags(self, *new_tags: Any) -> 'cirq.TaggedOperation':
+        """Creates a new TaggedOperation with combined tags.
+
+        Overloads Operation.with_tags to create a new TaggedOperation
+        that has the tags of this operation combined with the new_tags
+        specified as the parameter.
+        """
+        return TaggedOperation(self.sub_operation, *self._tags, *new_tags)
+
+    def __str__(self):
+        tag_repr = ','.join(repr(t) for t in self._tags)
+        return f"cirq.TaggedOperation({repr(self.sub_operation)}, {tag_repr})"
+
+    def __repr__(self):
+        return str(self)
+
+    def _value_equality_values_(self):
+        return (self.sub_operation, *self._tags)
```

Comment:
Why would it be less efficient than just the sub_operation plus tags?

Reply:
I was thinking to avoid creating a new tuple of length `1+len(tags)` by nesting tuples instead of flattening.  Now that I think about it, there won't usually be a large number of tags so it really doesn't matter.
```
# in __init__
self._tags = tuple(...)

def _value_equality_values_(self):
    return (self.sub_operation, self._tags)
```

Reply:
Ah, I was also confused about what you were concerned about.  I have changed it to a tuple, though I doubt the performance will matter too much.